include/pybind11/pybind11.h

```diff
@@ -44,27 +44,27 @@ class cpp_function : public function {
 
     /// Construct a cpp_function from a vanilla function pointer
     template <typename Return, typename... Args, typename... Extra>
-    cpp_function(Return (*f)(Args...), const Extra&... extra) {
+    explicit cpp_function(Return (*f)(Args...), const Extra&... extra) {
         initialize(f, f, extra...);
     }
 
     /// Construct a cpp_function from a lambda function (possibly with internal state)
-    template <typename Func, typename... Extra> cpp_function(Func &&f, const Extra&... extra) {
+    template <typename Func, typename... Extra> explicit cpp_function(Func &&f, const Extra&... extra) {
         initialize(std::forward<Func>(f),
                    (typename detail::remove_class<decltype(
                        &std::remove_reference<Func>::type::operator())>::type *) nullptr, extra...);
     }
 
     /// Construct a cpp_function from a class method (non-const)
     template <typename Return, typename Class, typename... Arg, typename... Extra>
-            cpp_function(Return (Class::*f)(Arg...), const Extra&... extra) {
+            explicit cpp_function(Return (Class::*f)(Arg...), const Extra&... extra) {
         initialize([f](Class *c, Arg... args) -> Return { return (c->*f)(args...); },
                    (Return (*) (Class *, Arg...)) nullptr, extra...);
     }
 
     /// Construct a cpp_function from a class method (const)
     template <typename Return, typename Class, typename... Arg, typename... Extra>
-            cpp_function(Return (Class::*f)(Arg...) const, const Extra&... extra) {
+            explicit cpp_function(Return (Class::*f)(Arg...) const, const Extra&... extra) {
```

Comment:
Turns out `cpp_function` needs to implicit. `explicit` breaks `def_property` and friends:
```c++
.def_property_readonly("value", &Foo::get_value) // compiler error: no viable conversion
```
Just caught this in my code. Looks like the tests don't cover any of the `def_property` variants.

The implicit conversion should be OK in general. This seems natural:
```c++
py::cpp_function = []() {
    // ...
};
```